File: cmd/risu-bridge/codex.go
package main

import (
	"bufio"
	"context"
	"encoding/json"
	"io"
	"os/exec"
	"sync"
	"time"
)

type rpcMessage struct {
	ID     json.RawMessage `json:"id"`
	Method string          `json:"method"`
	Params json.RawMessage `json:"params"`
	Result json.RawMessage `json:"result"`
	Error  json.RawMessage `json:"error"`
}
type codex struct {
	cmd         *exec.Cmd
	in          io.WriteCloser
	mu, writeMu sync.Mutex
	seq         int
	pending     map[string]chan rpcMessage
	events      map[string]chan rpcMessage
	done        chan struct{}
	once        sync.Once
}

func startCodex(binary, cwd string, env []string) (*codex, error) {
	c := &codex{pending: map[string]chan rpcMessage{}, events: map[string]chan rpcMessage{}, done: make(chan struct{})}
	c.cmd = exec.Command(binary, "app-server", "--listen", "stdio://")
	c.cmd.Dir = cwd
	c.cmd.Env = env
	c.cmd.Stderr = io.Discard
	var err error
	c.in, err = c.cmd.StdinPipe()
	if err != nil {
		return nil, err
	}
	out, err := c.cmd.StdoutPipe()
	if err != nil {
		return nil, err
	}
	if err = c.cmd.Start(); err != nil {
		return nil, err
	}
	go func() {
		s := bufio.NewScanner(out)
		s.Buffer(make([]byte, 65536), 8*1024*1024)
		for s.Scan() {
			var m rpcMessage
			if json.Unmarshal(s.Bytes(), &m) != nil {
				continue
			}
			if len(m.ID) > 0 {
				if m.Method != "" {
					_ = c.send(map[string]any{"id": m.ID, "error": map[string]any{"code": -32601, "message": "Interactive tools are disabled by this client."}})
					continue
				}
				c.mu.Lock()
				ch := c.pending[string(m.ID)]
				c.mu.Unlock()
				if ch != nil {
					select {
					case ch <- m:
					default:
					}
				}
			} else if m.Method != "" {
				var p struct {
					ThreadID string `json:"threadId"`
				}
				_ = json.Unmarshal(m.Params, &p)
				c.mu.Lock()
				ch := c.events[p.ThreadID]
				c.mu.Unlock()
				if ch != nil {
					select {
					case ch <- m:
					case <-c.done:
						return
					default:
						c.shutdown()
						return
					}
				}
			}
		}
		c.shutdown()
	}()
	go func() { _ = c.cmd.Wait(); c.once.Do(func() { close(c.done) }) }()
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()
	if err = c.rpc(ctx, "initialize", map[string]any{"clientInfo": map[string]string{"name": "risu_subscription_bridge", "title": "Risu Subscription Bridge", "version": version}, "capabilities": map[string]bool{"experimentalApi": true}}, nil); err != nil {
		c.shutdown()
		return nil, err
	}
	_ = c.send(map[string]any{"method": "initialized", "params": map[string]any{}})
	return c, nil
}
func (c *codex) send(v any) error {
	c.writeMu.Lock()
	defer c.writeMu.Unlock()
	_, err := c.in.Write(append(marshal(v), '\n'))
	return err
}
func (c *codex) rpc(ctx context.Context, method string, params, out any) error {
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	c.mu.Lock()
	c.seq++
	id := c.seq
	key := string(marshal(id))
	ch := make(chan rpcMessage, 1)
	c.pending[key] = ch
	c.mu.Unlock()
	defer func() { c.mu.Lock(); delete(c.pending, key); c.mu.Unlock() }()
	if err := c.send(map[string]any{"id": id, "method": method, "params": params}); err != nil {
		return problem(503, "harness_stopped", "Codex is unavailable. Restart the bridge.")
	}
	select {
	case m := <-ch:
		if len(m.Error) > 0 && string(m.Error) != "null" {
			return problem(502, "rpc_error", "Codex RPC failed.")
		}
		if out != nil {
			return json.Unmarshal(m.Result, out)
		}
		return nil
	case <-ctx.Done():
		return ctx.Err()
	case <-c.done:
		return problem(503, "harness_stopped", "Codex stopped. Restart the bridge.")
	}
}
func (c *codex) shutdown() {
	c.once.Do(func() {
		close(c.done)
		_ = c.in.Close()
		if c.cmd.Process != nil {
			_ = c.cmd.Process.Kill()
		}
	})
}
func (c *codex) alive() bool {
	select {
	case <-c.done:
		return false
	default:
		return true
	}
}

type account struct {
	Connected bool   `json:"connected"`
	Type      string `json:"type"`
	Plan      string `json:"plan"`
}

func (c *codex) account(ctx context.Context) (account, error) {
	var r struct {
		Account *struct {
			Type string `json:"type"`
			Plan string `json:"planType"`
		} `json:"account"`
	}
	err := c.rpc(ctx, "account/read", map[string]bool{"refreshToken": false}, &r)
	a := account{}
	if r.Account != nil {
		a = account{r.Account.Type == "chatgpt", r.Account.Type, r.Account.Plan}
	}
	return a, err
}

type model struct {
	ID          string `json:"id"`
	Model       string `json:"model"`
	DisplayName string `json:"displayName"`
	IsDefault   bool   `json:"isDefault"`
	Efforts     []struct {
		Effort string `json:"reasoningEffort"`
	} `json:"supportedReasoningEfforts"`
}

func (m model) name() string {
	if m.Model != "" {
		return m.Model
	}
	return m.ID
}
func (c *codex) models(ctx context.Context) ([]model, error) {
	ms := []model{}
	cursor := ""
	for {
		p := map[string]any{"limit": 100}
		if cursor != "" {
			p["cursor"] = cursor
		}
		var r struct {
			Data       []model `json:"data"`
			NextCursor string  `json:"nextCursor"`
		}
		if err := c.rpc(ctx, "model/list", p, &r); err != nil {
			return nil, err
		}
		ms = append(ms, r.Data...)
		cursor = r.NextCursor
		if cursor == "" {
			return ms, nil
		}
	}
}
func chooseModel(ms []model, name, effort string) (model, error) {
	var selected *model
	for i := range ms {
		if name == "subscription-default" {
			if selected == nil || ms[i].IsDefault {
				selected = &ms[i]
			}
		} else if ms[i].name() == name || ms[i].ID == name {
			selected = &ms[i]
			break
		}
	}
	if selected == nil {
		return model{}, problem(400, "unknown_model", "Choose an available model.")
	}
	if effort != "" {
		ok := false
		for _, e := range selected.Efforts {
			if e.Effort == effort {
				ok = true
			}
		}
		if !ok {
			return model{}, problem(400, "effort_invalid", "선택 모델이 추론 강도를 지원하지 않습니다. 기본값으로 바꾸세요.")
		}
	}
	return *selected, nil
}

type generationResult struct {
	Model        string           `json:"model"`
	Usage        map[string]int64 `json:"usage"`
	FirstTokenMS *int64           `json:"firstTokenMs"`
	ElapsedMS    int64            `json:"elapsedMs"`
}

func (c *codex) generate(ctx context.Context, r chatRequest, s settings, cwd string, delta func(string)) (result generationResult, err error) {
	start := time.Now()
	ctx, cancel := context.WithTimeout(ctx, 180*time.Second)
	defer cancel()
	a, err := c.account(ctx)
	if err != nil {
		return result, err
	}
	if !a.Connected {
		return result, problem(401, "login_required", "Sign in with ChatGPT on the local setup page.")
	}
	ms, err := c.models(ctx)
	if err != nil {
		return result, err
	}
	name := r.Model
	if name == "subscription-default" {
		name = s.Model
	}
	m, err := chooseModel(ms, name, s.Effort)
	if err != nil {
		return result, err
	}
	result.Model = m.name()
	p := map[string]any{"model": result.Model, "cwd": cwd, "ephemeral": true, "sandbox": "read-only", "approvalPolicy": "never", "baseInstructions": baseInstructions}
	if s.Instructions != "" {
		p["baseInstructions"] = baseInstructions + "\n\n" + s.Instructions
	}
	if s.Verbosity != "" {
		p["config"] = map[string]string{"model_verbosity": s.Verbosity}
	}
	var thread struct {
		Thread struct {
			ID string `json:"id"`
		} `json:"thread"`
	}
	if err = c.rpc(ctx, "thread/start", p, &thread); err != nil {
		return result, err
	}
	tid := thread.Thread.ID
	if tid == "" {
		return result, problem(502, "rpc_error", "Missing thread ID.")
	}
	ch := make(chan rpcMessage, 2048)
	c.mu.Lock()
	c.events[tid] = ch
	c.mu.Unlock()
	turnID := ""
	defer func() {
		if err != nil && turnID != "" {
			clean, done := context.WithTimeout(context.Background(), 5*time.Second)
			_ = c.rpc(clean, "turn/interrupt", map[string]string{"threadId": tid, "turnId": turnID}, nil)
			done()
		}
		clean, done := context.WithTimeout(context.Background(), 5*time.Second)
		defer done()
		if e := c.rpc(clean, "thread/unsubscribe", map[string]string{"threadId": tid}, nil); e != nil {
			c.shutdown()
			err = problem(503, "cleanup_failed", "Thread cleanup failed. Restart the bridge.")
		}
		c.mu.Lock()
		delete(c.events, tid)
		c.mu.Unlock()
	}()
	p = map[string]any{"threadId": tid, "input": []any{map[string]any{"type": "text", "text": promptFor(r.Messages), "text_elements": []any{}}}}
	if s.Effort != "" {
		p["effort"] = s.Effort
	}
	var turn struct {
		Turn struct {
			ID string `json:"id"`
		} `json:"turn"`
	}
	// Keep start bounded independently of the client so its returned turn ID is available for cancellation.
	startCtx, finish := context.WithTimeout(context.Background(), 30*time.Second)
	err = c.rpc(startCtx, "turn/start", p, &turn)
	finish()
	turnID = turn.Turn.ID
	if err != nil {
		return result, err
	}
	for {
		select {
		case <-ctx.Done():
			return result, ctx.Err()
		case <-c.done:
			return result, problem(503, "harness_stopped", "Codex stopped during generation.")
		case event := <-ch:
			var p struct {
				Delta string `json:"delta"`
				Turn  struct {
					ID, Status string
					Error      json.RawMessage
				} `json:"turn"`
				TokenUsage struct {
					Last *struct{ InputTokens, OutputTokens, TotalTokens int64 } `json:"last"`
				} `json:"tokenUsage"`
			}
			if json.Unmarshal(event.Params, &p) != nil {
				continue
			}
			switch event.Method {
			case "turn/started":
				turnID = p.Turn.ID
			case "item/agentMessage/delta":
				if result.FirstTokenMS == nil {
					n := time.Since(start).Milliseconds()
					result.FirstTokenMS = &n
				}
				delta(p.Delta)
			case "thread/tokenUsage/updated":
				if u := p.TokenUsage.Last; u != nil {
					result.Usage = map[string]int64{"prompt_tokens": u.InputTokens, "completion_tokens": u.OutputTokens, "total_tokens": u.TotalTokens}
				}
			case "turn/completed":
				if p.Turn.Status != "completed" {
					var e struct {
						Info json.RawMessage `json:"codexErrorInfo"`
					}
					_ = json.Unmarshal(p.Turn.Error, &e)
					if isQuota(string(e.Info)) {
						return result, problem(429, "rate_limit", "Subscription limit reached. Wait for reset.")
					}
					return result, problem(502, "generation_failed", "Generation failed.")
				}
				result.ElapsedMS = time.Since(start).Milliseconds()
				return result, nil
			}
		}
	}
}

File: src/codex.mjs
import {spawn} from 'node:child_process';
import {createInterface} from 'node:readline';
import {EventEmitter} from 'node:events';
import {BridgeError, BASE, promptFor} from './protocol.mjs';

export class Codex extends EventEmitter {
  constructor({binary, env, cwd}) { super(); Object.assign(this, {binary, env, cwd}); this.pending = new Map(); this.seq = 0; }
  async init() {
    this.child = spawn(this.binary, ['app-server', '--listen', 'stdio://'], {env: this.env, cwd: this.cwd, stdio: ['pipe', 'pipe', 'pipe']});
    // Never forward raw harness logs: they may include transcript fragments or credentials.
    this.child.stderr.resume();
    this.child.stdin.on('error', () => {});
    createInterface({input: this.child.stdout}).on('line', line => {
      let m; try { m = JSON.parse(line); } catch { return; }
      if (m.id !== undefined && m.method) {
        // No model-initiated approvals, tool calls, or interactions are allowed.
        this.send({id: m.id, error: {code: -32601, message: 'Interactive tools are disabled by this client.'}});
      } else if (m.id !== undefined) {
        const p = this.pending.get(m.id); if (!p) return;
        clearTimeout(p.timer); this.pending.delete(m.id);
        m.error ? p.reject(new BridgeError(`Codex RPC failed (${m.error.code ?? 'unknown'}).`, 502, 'rpc_error')) : p.resolve(m.result);
      } else if (m.method) this.emit('notification', m);
    });
    const fail = () => {
      this.alive = false;
      for (const p of this.pending.values()) { clearTimeout(p.timer); p.reject(new BridgeError('Codex stopped. Restart the bridge.', 503, 'harness_stopped')); }
      this.pending.clear(); this.emit('stopped');
    };
    this.child.once('error', fail); this.child.once('exit', fail); this.alive = true;
    await this.rpc('initialize', {clientInfo: {name: 'risu_subscription_bridge', title: 'Risu Subscription Bridge', version: '0.1.0'}, capabilities: {experimentalApi: true}});
    this.send({method: 'initialized', params: {}});
  }
  send(m) { if (this.child?.stdin.writable) this.child.stdin.write(JSON.stringify(m) + '\n'); }
  rpc(method, params = {}, timeout = 30000) {
    if (!this.alive) return Promise.reject(new BridgeError('Codex is unavailable.', 503, 'harness_stopped'));
    const id = ++this.seq;
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => { this.pending.delete(id); reject(new BridgeError('Codex request timed out.', 504, 'rpc_timeout')); }, timeout);
      this.pending.set(id, {resolve, reject, timer}); this.send({id, method, params});
    });
  }
  async account() {
    const {account} = await this.rpc('account/read', {refreshToken: false});
    return {connected: account?.type === 'chatgpt', type: account?.type ?? null, plan: account?.planType ?? null};
  }
  login() { return this.rpc('account/login/start', {type: 'chatgpt'}); }
  async models() {
    let cursor; const models = [];
    do { const r = await this.rpc('model/list', {limit: 100, ...(cursor ? {cursor} : {})}); models.push(...r.data); cursor = r.nextCursor; } while (cursor);
    return models;
  }
  async generate(request, {signal, delta}) {
    if (signal.aborted) throw new BridgeError('Cancelled.', 499, 'cancelled');
    const account = await this.account();
    if (!account.connected) throw new BridgeError('Sign in with ChatGPT on the local setup page. API-key accounts are not accepted.', 401, 'login_required');
    const available = await this.models();
    const chosen = request.model === 'subscription-default' ? available.find(m => m.isDefault) ?? available[0] : available.find(m => m.model === request.model || m.id === request.model);
    if (!chosen) throw new BridgeError('Choose a model returned by /v1/models.', 400, 'unknown_model');
    const model = chosen.model ?? chosen.id;
    const {thread} = await this.rpc('thread/start', {model, cwd: this.cwd, ephemeral: true, sandbox: 'read-only', approvalPolicy: 'never', baseInstructions: BASE});
    let turnId, timer, usage, settled = false, startedAt = performance.now(), firstTokenMs = null;
    let resolveDone, rejectDone;
    const done = new Promise((resolve, reject) => { resolveDone = resolve; rejectDone = reject; });
    // Attach a rejection handler immediately while turn/start is still in flight.
    done.catch(() => {});
    const settle = error => { if (settled) return; settled = true; error ? rejectDone(error) : resolveDone(); };
    const interrupt = () => { if (turnId) this.rpc('turn/interrupt', {threadId: thread.id, turnId}, 5000).catch(() => {}); };
    const abort = () => { interrupt(); settle(new BridgeError('Cancelled.', 499, 'cancelled')); };
    const stopped = () => settle(new BridgeError('Codex stopped during generation.', 503, 'harness_stopped'));
    const listen = ({method, params: p}) => {
      if (p?.threadId !== thread.id || settled) return;
      if (method === 'turn/started') { turnId = p.turn.id; if (signal.aborted) interrupt(); }
      if (method === 'item/agentMessage/delta') { firstTokenMs ??= Math.round(performance.now() - startedAt); delta(p.delta); }
      if (method === 'thread/tokenUsage/updated') {
        const u = p.tokenUsage?.last;
        if (u) usage = {prompt_tokens: u.inputTokens, completion_tokens: u.outputTokens, total_tokens: u.totalTokens};
      }
      if (method === 'turn/completed') {
        turnId ??= p.turn.id;
        if (p.turn.status === 'completed') settle();
        else {
          const limit = /limit|quota|usage/i.test(JSON.stringify(p.turn.error?.codexErrorInfo ?? ''));
          settle(new BridgeError(limit ? 'Subscription limit reached. Wait for reset.' : `Generation ${p.turn.status}.`, limit ? 429 : 502, limit ? 'rate_limit' : 'generation_failed'));
        }
      }
    };
    this.on('notification', listen); this.once('stopped', stopped); signal.addEventListener('abort', abort, {once: true});
    timer = setTimeout(() => { interrupt(); settle(new BridgeError('Generation exceeded 180 seconds.', 504, 'generation_timeout')); }, 180000);
    try {
      if (signal.aborted) abort();
      else {
        const r = await this.rpc('turn/start', {threadId: thread.id, input: [{type: 'text', text: promptFor(request.messages), text_elements: []}]});
        turnId = r.turn.id; if (signal.aborted || settled && r.turn.status === 'inProgress') interrupt();
      }
      await done;
      return {model, usage, firstTokenMs, elapsedMs: Math.round(performance.now() - startedAt)};
    } finally {
      clearTimeout(timer); this.off('notification', listen); this.off('stopped', stopped); signal.removeEventListener('abort', abort);
      await this.rpc('thread/unload', {threadId: thread.id}, 5000).catch(() => {});
    }
  }
  shutdown() { this.child?.stdin.end(); const child = this.child; const t = setTimeout(() => child?.kill('SIGTERM'), 1500); t.unref(); }
}
